zerver/lib/streams.py

```diff
@@ -1163,11 +1172,45 @@ def filter_stream_authorization(
         stream
         for stream in streams
         if stream.id not in {stream.id for stream in unauthorized_streams}
-        and stream.id not in {stream.id for stream in streams_to_which_user_cannot_add_subscribers}
     ]
-    return StreamsCategorizedByPermissions(
+    return BaseStreamsCategorizedByPermissions(
         authorized_streams=authorized_streams,
         unauthorized_streams=unauthorized_streams,
+    )
+
+
+def filter_stream_authorization_for_adding_subscriptions(
+    user_profile: UserProfile, streams: Collection[Stream], is_subscribing_other_users: bool = False
+) -> StreamsCategorizedByPermissionsForAddingSubscribers:
+    if len(streams) == 0:
+        return StreamsCategorizedByPermissionsForAddingSubscribers(
+            authorized_streams=[],
+            unauthorized_streams=[],
+            streams_to_which_user_cannot_add_subscribers=[],
+        )
+
+    user_group_membership_details = UserGroupMembershipDetails(user_recursive_group_ids=None)
+    filter_stream_authorization_result = filter_stream_authorization(
+        user_profile, streams, user_group_membership_details
+    )
+
+    streams_to_which_user_cannot_add_subscribers: list[Stream] = []
+    if is_subscribing_other_users:
+        streams_to_which_user_cannot_add_subscribers = (
+            get_streams_to_which_user_cannot_add_subscribers(
+                list(filter_stream_authorization_result.authorized_streams),
+                user_profile,
+                user_group_membership_details=user_group_membership_details,
+            )
+        )
+    authorized_streams = [
+        stream
+        for stream in filter_stream_authorization_result.authorized_streams
+        if stream.id not in {stream.id for stream in streams_to_which_user_cannot_add_subscribers}
```

Comment:
I think this constructs the set of `{stream.id for stream in streams_to_which_user_cannot_add_subscribers}` on each iteration through the loop. Just declare a set with a variable before the comprehension.